lib/barcelona/plugins/itamae_plugin.rb

```diff
@@ -0,0 +1,73 @@
+# Itamae plugin
+# This plugin installs Itamae and apply the specified recipe on initialize
+# Usage: bcn district put-plugin essa-test itamae -a recipe_url=s3://barcelona-essa-test-1573458106/itamae_recipes/recipe.tar.gz
+
+module Barcelona
+  module Plugins
+    class ItamaePlugin < Base
+      def on_container_instance_user_data(_instance, user_data)
+        install_itamae(user_data)
+        apply_itamae_recipe(user_data, attributes['recipe_url'], 'barcelona_container')
+        user_data
+      end
+
+      def on_network_stack_template(_stack, template)
+        bastion_lc = template["BastionLaunchConfiguration"]
+        return template if bastion_lc.nil?
+
+        user_data = InstanceUserData.load_or_initialize(bastion_lc["Properties"]["UserData"])
+        install_itamae(user_data)
+        apply_itamae_recipe(user_data, attributes['recipe_url'], 'barcelona_bastion')
+        bastion_lc["Properties"]["UserData"] = user_data.build
+        template
+      end
+
+      private
+      def install_itamae(user_data)
+        user_data.packages += %w(
+          gcc-c++
+          redhat-rpm-config
+        )
+        user_data.run_commands += <<~EOS.split("\n")
+          amazon-linux-extras install ruby2.4
+          yum install -y ruby-devel
+          gem install itamae io-console -N
+        EOS
+      end
+
+      def apply_itamae_recipe(user_data, recipe_url, role)
+        archive = Pathname.new(URI.parse(recipe_url).path).basename
+        user_data.add_file("/usr/local/bin/apply_itamae.sh", "root:root", "755", <<~EOS)
+          #!/bin/bash
+          set -ex
+          # download the recipe
+          mkdir /root/itamae && cd /root/itamae
+          aws s3 cp #{recipe_url} .
+          tar zxvf #{archive}
+          # Make sudo usable without tty
+          sed -i -e "s/^ *Defaults *requiretty/# Defaults requiretty/" /etc/sudoers
+          # apply itamae recipe
+          itamae local roles/#{role}.rb -y nodes/#{role}.yml -l info > /var/log/itamae.log
+        EOS
+
+        user_data.add_file("/etc/systemd/system/itamae.service", "root:root", "644", <<~EOS)
```

Comment:
Why is systemd used for executing itamae? is it not possible to just execute it inside the usedata?

Reply:
There was an issue on redhat. Itamae uses `su` many times even if invoked on `root` and `su` fails on cloud-init on redhat. Invoking as a systemd service is a workaround for it.

This issue didn't happen on Amazon Linux2 but I used the same method for safety and uniformity. Because I couldn't figure out exact reason why `su` fails in such case.

I'm still wondering which to choose. How do you think about it?

Reply:
Hmm that is strange. I'd like to keep this script simple so if it works on amazon linux I want to choose the one-line solution.
Did the error occur every time on RedHat but it doesn't at all on Amazon Linux? If so, let's just try testing on staging several times to make sure that it is not an issue for Amazon Linux

Reply:
Thanks for the comment.

This issue happens only on RedHat, not on Amazon Linux2. So, I will remove the indirect invocation of Itamae.

Actually, `su` didn't fail but became very slow. I can find many similar issues in Internet.

* [Linuxのsuやsshが遅い \- 銀の鍵 \(The Silver Key\)](https://www.ginnokagi.com/2014/05/linuxsussh.html)

I think my issue is related with pam or name resolution on authentication as this post describes. But editing `/etc/hosts` didn't work.

I can watch the slow `su` while initialization continues. But `su` on my ssh was not slow at the same time. This means that I need to edit `/etc/hosts` or other configuration **before** `runcmd` starts to identify the exact cause.

So I abandoned investigation and used indirect invocation on RedHat.